File: backend/examples/kotlin/RunCorpus.kt
package io.heimui.hydration

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.io.File
import kotlin.system.exitProcess

/**
 * Runs the published hydration corpus against this implementation.
 *
 * Every case is a screen, a payload, the document a device should receive, and the expressions that
 * should have been reported as unresolved. If all of them pass, this implementation agrees with
 * every other one -- which is the only definition of correct that means anything here.
 *
 *     kotlin RunCorpusKt [path/to/corpus]
 */
fun main(args: Array<String>) {
    val json = Json { ignoreUnknownKeys = true }
    val dir = File(args.firstOrNull() ?: "../../corpus")
    val cases = dir.listFiles { f -> f.extension == "json" }?.sortedBy { it.name }.orEmpty()
    if (cases.isEmpty()) {
        System.err.println("no cases found in ${dir.absolutePath}")
        exitProcess(1)
    }

    val failures = mutableListOf<String>()
    for (file in cases) {
        val case = json.parseToJsonElement(file.readText()).jsonObject
        val policy = when (case["options"]?.jsonObject?.get("onUnresolved")?.jsonPrimitive?.contentOrNull) {
            "blank" -> UnresolvedPolicy.BLANK
            else -> UnresolvedPolicy.KEEP
        }
        val result = HeimHydrationEngine.hydrateWithReport(case["screen"]!!.jsonObject, case["data"]!!.jsonObject, policy)

        val before = failures.size
        if (result.document != case["expected"]!!.jsonObject) {
            failures += "${file.name}: document\n  expected ${case["expected"]}\n  but was  ${result.document}"
        }
        val wanted = (case["expectedUnresolved"] as? JsonArray).orEmpty().map {
            val o = it.jsonObject
            "${o["nodeId"]!!.jsonPrimitive.content}|${o["property"]!!.jsonPrimitive.content}|${o["expression"]!!.jsonPrimitive.content}"
        }
        val got = result.unresolved.map { "${it.nodeId}|${it.property}|${it.expression}" }
        if (wanted != got) failures += "${file.name}: unresolved\n  expected $wanted\n  but was  $got"

        println(if (failures.size == before) "  ok   ${file.name}" else "  FAIL ${file.name}")
    }

    println()
    if (failures.isNotEmpty()) {
        System.err.println(failures.joinToString("\n"))
        exitProcess(1)
    }
    println("${cases.size}/${cases.size} cases pass.")
}

private fun JsonArray?.orEmpty(): List<kotlinx.serialization.json.JsonElement> = this ?: emptyList()

File: backend/examples/python/run_corpus.py
"""
Runs the published hydration corpus against this implementation.

Every case is a screen, a payload, the document a device should receive, and the expressions that
should have been reported as unresolved. If all of them pass, this implementation agrees with every
other one -- which is the only definition of correct that means anything here.

    python3 run_corpus.py [path/to/corpus]
"""
import json
import pathlib
import sys

from heimui_hydration import UnresolvedPolicy, hydrate_with_report

corpus = pathlib.Path(sys.argv[1] if len(sys.argv) > 1 else "../../corpus")
cases = sorted(corpus.glob("*.json"))
if not cases:
    sys.exit(f"no cases found in {corpus}")

failures = []
for path in cases:
    case = json.loads(path.read_text())
    policy = UnresolvedPolicy.BLANK if case.get("options", {}).get("onUnresolved") == "blank" else UnresolvedPolicy.KEEP
    result = hydrate_with_report(case["screen"], case["data"], policy)

    if result.document != case["expected"]:
        failures.append(
            f"{path.name}: document\n  expected {json.dumps(case['expected'], sort_keys=True)}\n"
            f"  but was  {json.dumps(result.document, sort_keys=True)}"
        )

    wanted = [(u["nodeId"], u["property"], u["expression"]) for u in case.get("expectedUnresolved", [])]
    got = [(u.node_id, u.property, u.expression) for u in result.unresolved]
    if wanted != got:
        failures.append(f"{path.name}: unresolved\n  expected {wanted}\n  but was  {got}")

    print(("  ok   " if not failures or failures[-1].split(':')[0] != path.name else "  FAIL ") + path.name)

print()
if failures:
    print("\n".join(failures))
    sys.exit(f"{len(failures)} failure(s) across {len(cases)} cases")
print(f"{len(cases)}/{len(cases)} cases pass.")
